src/cleaner/scripts/clean_docker.py

```diff
@@ -0,0 +1,85 @@
+# Copyright (c) Microsoft Corporation
+# All rights reserved.
+#
+# MIT License
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+# to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+# BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+from cleaner.utils import common
+import subprocess
+import multiprocessing
+import re
+
+logger = multiprocessing.get_logger()
+
+
+def check_disk_usage(partition):
+    df = subprocess.Popen(["df","-h", partition], stdout=subprocess.PIPE)
+    size = 0
+    try:
+        for line in df.stdout:
+            splitline = line.decode().split()
+            if splitline[5] == partition:
+                size = int(splitline[4][:-1])
+    except ValueError:
+        logger.error("cannot get disk size, reset size to 0")
+        size = 0
+    logger.info("Checking disk, disk usage = {0}%".format(size))
+    return size
+
+
+def check_and_clean(threshold):
+    if check_disk_usage("/") > int(threshold):
+        logger.info("Disk usage is above {0}%, Try to remove containers".format(threshold))
+        kill_largest_container()
+
+
+# Clean logic v1: kill largest container
+white_list = ["k8s_kube", "k8s_pylon", "k8s_zookeeper", "k8s_rest-server", "k8s_yarn", "k8s_hadoop", "k8s_job-exporter", "k8s_watchdog", "k8s_grafana", "k8s_node-exporter", "k8s_webportal", "k8s_prometheus", "k8s_nvidia-drivers", "k8s_etcd-container", "k8s_apiserver-container", "k8s_docker-cleaner", "kubelet"]
+def kill_largest_container():
+    containers = []
+    # Only try to stop PAI jobs and user created containers
+    containers_source = subprocess.Popen(["docker", "ps", "-a", "--format", r'{{.ID}}\t{{.Image}}\t{{.Size}}\t{{.Names}}'], stdout=subprocess.PIPE)
```

Comment:
We might need to record what size is considered in `docker ps`.
Currently following items is excluded:
- disk space used for log-files (if you use the json-file logging driver) - which can be quite a bit if your container generates a lot of logs, and log-rotation (max-file / max-size logging options) is not configured
- volumes used by the container
- disk space used for the container's configuration files (hostconfig.json, config.v2.json, hosts, hostname, resolv.conf) - although these files are small
- memory written to disk (if swapping is enabled)
- checkpoints (if you're using the experimental checkpoint/restore feature)

Comes from https://github.com/docker/docker.github.io/issues/1520

Reply:
Not include in this version.